### PR TITLE
[14.0][ADD] l10n_it_delivery_note_rma

### PR DESCRIPTION
**Pull request by solo4games**
This module adds:
- the possibility to create one or more RMA(s) from a Delivery Note (from l10n_it_delivery_note, Italian legally required transportation document)
- a link between DN and RMA(s)
- adding an "RMA" smart button in DN to quickly access RMAs created from a specific DN

### CODE REVIEW

File: l10n_it_delivery_note_rma/models/rma.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    @api.depends("delivery_note_id")
+    def _compute_allowed_product_ids(self):
+        for rec in self:
+            if rec.delivery_note_id:
+                delvery_note_product = rec.delivery_note_id.line_ids.mapped(
+                    "product_id"
+                )
+                rec.allowed_product_ids = delvery_note_product.filtered(
+                    lambda r: r.type in ["consu", "product"]
+                ).ids
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/rma.py:76]: Why do you use ids property?

File: l10n_it_delivery_note_rma/models/rma.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            else:
+                rec.allowed_product_ids = False  # don't populate a big list
+
+    @api.onchange("partner_id")
+    def _onchange_partner_id(self):
+        res = super()._onchange_partner_id()
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/rma.py:82]: onchange function doesn't return value.
```
super()._onchange_partner_id()
self.delivery_note_id = False
```

File: l10n_it_delivery_note_rma/models/rma.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        if not self.delivery_note_line_id:
+            return super()._get_refund_line_product()
+        return self.delivery_note_line_id.product_id
+
+    def _prepare_refund_line(self, line_form):
+        """Add line data"""
+        super()._prepare_refund_line(line_form)
+        line = self.delivery_note_line_id
+        if line:
+            line_form.discount = line.discount
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/rma.py:115]: Here better solution will use `write()` function:
```
line_form.write({
    "discount": line.discount.
    "sequence": line.sequence,
})

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+
+class StockDeliveryNoteLine(models.Model):
+    _inherit = "stock.delivery.note.line"
+
+    delivery_move_ids = fields.One2many(
+        "stock.move", "delivery_note_line_id", string="Delivery Transfers"
+    )
+
+    def get_delivery_move(self):
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:126]: Why did you create this function?

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    def _get_invoiced(self):
+        """Search for possible RMA refunds and link them to the delivery note. We
+        don't want to link their delivery note lines as that would unbalance the
+        qtys to invoice wich isn't correct for this case"""
+        super()._get_invoiced()
+        for order in self:
+            refunds = order.sudo().rma_ids.mapped("refund_id")
+            if not refunds:
+                continue
+            order.invoice_ids += refunds
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:115]: Use `write` function

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+    rma_count = fields.Integer(string="RMA count", compute="_compute_rma_count")
+
+    warehouse_id = fields.Many2one(
+        "stock.warehouse",
+        string="Warehouse",
+        required=True,
+        readonly=True,
+        states={"draft": [("readonly", False)], "sent": [("readonly", False)]},
+        default=_default_warehouse_id,
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:30]: Better solution will be that:
```
default=lambda self: self.env.user._get_default_warehouse_id()
```

File: l10n_it_delivery_note_rma/wizard/stock_delivery_note_rma_wizard.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    def create_rma(self):
+        self.ensure_one()
+        user_has_group_portal = self.env.user.has_group(
+            "base.group_portal"
+        ) or self.env.user.has_group("base.group_public")
+        lines = self.line_ids.filtered(lambda r: r.quantity > 0.0)
+        val_list = [line._prepare_rma_values() for line in lines]
+        rma_model = (
+            self.env["rma"].with_user(SUPERUSER_ID)
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/wizard/stock_delivery_note_rma_wizard.py:53]: For what do you use SUPERUSER?

**Review by geomer198 — CHANGES_REQUESTED** [review]
Would you add the tests for code.

**Comment by aayartsev** [comment]
You need to add tests. Other cases are normal

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            "res_model": "stock.delivery.note.rma.wizard",
+            "res_id": wizard.id,
+            "target": "new",
+        }
+
+    def action_view_rma(self):
+        self.ensure_one()
+        action = self.sudo().env.ref("rma.rma_action").read()[0]
+        rma = self.rma_ids
+        if len(rma) == 1:
```
Comment (aayartsev) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:86]: It is not clear, why you need to handle case when you have singletone?

**Review by aayartsev — CHANGES_REQUESTED** [review]

File: l10n_it_delivery_note_rma/tests/delivery_note_common.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+DOWNPAYMENT_METHODS = ["fixed", "percentage"]
+
+
+class StockDeliveryNoteCommon(TransactionCase):
+    def create_partner(self, name, **kwargs):
+        return self.env["res.partner"].create({"name": name, **kwargs})
+
+    def create_sales_order(self, lines, **kwargs):
+        vals = {"partner_id": self.recipient.id}
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/tests/delivery_note_common.py:11]: You can use Form for SO creation

**Review by geomer198 — COMMENTED** [review]
Add message for each assertion.
For Example:
```
self.assertFalse(rma.reception_move_id.origin_returned_move_id, msg="Move must be not set")
```

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class StockDeliveryNote(models.Model):
+    _inherit = "stock.delivery.note"
+
+    @api.model
+    def _default_warehouse_id(self):
```
Comment (geomer198) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:9]: This function don't used, remove please.

**Review by geomer198 — APPROVED** [review]
Code Review ok!

**Review by aayartsev — APPROVED** [review]

**Review by francesco-ooops — APPROVED** [review]
Review funzionale ok!

**Comment by francesco-ooops** [comment]
@OCA/local-italy-maintainers potete fare una review?

File: l10n_it_delivery_note_rma/models/stock_delivery_note.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            "type": "ir.actions.act_window",
+            "view_mode": "form",
+            "res_model": "stock.delivery.note.rma.wizard",
+            "res_id": wizard.id,
+            "target": "new",
+        }
+
+    def action_view_rma(self):
+        self.ensure_one()
+        action = self.sudo().env.ref("rma.rma_action").read()[0]
```
Comment (GSLabIt) [inline l10n_it_delivery_note_rma/models/stock_delivery_note.py:78]: what about:
```
action = self.env["ir.actions.act_window"]._for_xml_id("rma.rma_action")
```

**Review by GSLabIt — CHANGES_REQUESTED** [review]

File: l10n_it_delivery_note_rma/models/rma.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+class Rma(models.Model):
+    _inherit = "rma"
+
+    delivery_note_id = fields.Many2one(
+        comodel_name="stock.delivery.note",
+        string="Delivery Note",
+        domain="["
+        "    ('partner_id', 'child_of', commercial_partner_id),"
+        "    ('state', 'in', ['confirm', 'done', 'invoiced']),"
+        "]",
```
[the remark this answers is not in the document]
Comment (solo4games) [reply l10n_it_delivery_note_rma/models/rma.py:13]: There is no field 'delivery_note_id' in my inherited view, so how can I add domain there?

[the remark this answers is not in the document]
Comment (GSLabIt) [reply l10n_it_delivery_note_rma/models/rma.py:13]: > There is no field 'delivery_note_id' in my inherited view, so how can I add domain there?

if not in the view, why u need to add domain? u could filter when u need it ..isnt it?

[the remark this answers is not in the document]
Comment (solo4games) [reply l10n_it_delivery_note_rma/models/rma.py:13]: Field is in the original rma view, but in my module where I inherited this view, I don't have this field, that's why I can not use domain in the view, but maybe I am wrong.

File: oca_dependencies.txt [hunk 13]
```diff
@@ -1,2 +1,3 @@
 # See https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#oca_dependencies-txt
 web
+rma
```
Comment (francesco-ooops) [inline oca_dependencies.txt:3]: @solo4games you can remove this due to #3296

**Review by francesco-ooops — CHANGES_REQUESTED** [review]

File: oca_dependencies.txt (hunk 13 above)
Comment (solo4games) [reply oca_dependencies.txt]: Done, check pls

**Review by GSLabIt — APPROVED** [review]

**Review by francesco-ooops — APPROVED** [review]
@OCA/local-italy-maintainers pronta per il merge?

**Comment by francesco-ooops** [comment]
@OCA/local-italy-maintainers potete fare review? grazie!

**Comment by francesco-ooops** [comment]
@OCA/local-italy-maintainers 4 approvazioni, possibile mergiare?

File: l10n_it_delivery_note_rma/__manifest__.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "Return Merchandise Authorization Management - Link with Delivery Notes",
```
Comment (tafaRU) [inline l10n_it_delivery_note_rma/__manifest__.py:2]: Please adapt it to the following convention: https://www.odoo-italia.org/documentazione/14.0/sviluppo/guidelines.html#titolo-del-modulo

**Review by tafaRU — CHANGES_REQUESTED** [review]
Please also squash all the commits to only one.
The rest LGTM, thanks!

File: l10n_it_delivery_note_rma/__manifest__.py (hunk 14 above)
Comment (francesco-ooops) [reply l10n_it_delivery_note_rma/__manifest__.py:2]: @tafaRU può andare "ITA - Creazione RMA da DDT (Delivery Note)" ?

Comment (tafaRU) [reply l10n_it_delivery_note_rma/__manifest__.py:2]: :+1:

Comment (solo4games) [reply l10n_it_delivery_note_rma/__manifest__.py:2]: Ok

**Comment by francesco-ooops** [comment]
@tafaRU ok ora? :)

**Pull request merged**